File: Python/Trapping_Rain_Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        ans=0
        maxl=maxr=0
        l=0
        r=len(height)-1
        while(l<r):
            if(height[l]<height[r]):
                if height[l]>maxl:
                    maxl=height[l]
                else:
                    ans+=maxl-height[l]
                l+=1
            else:
                if height[r]>maxr:
                    maxr=height[r]
                else:
                    ans+=maxr-height[r]
                r-=1
        return ans


           
                
                
                    
                    
                    
                    
                    
                    
        
        
        

File: C++/trapping_rain_water.cpp
/*
Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.
*/
class Solution {
public:
    int trap(vector<int>& height) {
        int l=0;
        int r=height.size()-1;
        int level=0;
        int water=0;
        while(l<r)
        {
            int lower=height[height[l]<height[r]?l++:r--];
            level=max(level,lower);
            water +=level - lower;
        }
        return water;
    }
};